plugin/pkg/scheduler/schedulercache/node_info.go

```diff
@@ -68,29 +72,84 @@ func (n *NodeInfo) RequestedResource() Resource {
 	return *n.requestedResource
 }
 
+// NonZeroRequest returns aggregated nonzero resource request of pods on this node.
+func (n *NodeInfo) NonZeroRequest() Resource {
+	if n == nil {
+		return emptyResource
+	}
+	return *n.nonzeroRequest
+}
+
+func (n *NodeInfo) Clone() *NodeInfo {
+	pods := append([]*api.Pod(nil), n.pods...)
+	clone := &NodeInfo{
+		requestedResource: &(*n.requestedResource),
+		nonzeroRequest:    &(*n.nonzeroRequest),
+		pods:              pods,
+	}
+	return clone
+}
+
 // String returns representation of human readable format of this NodeInfo.
 func (n *NodeInfo) String() string {
 	podKeys := make([]string, len(n.pods))
 	for i, pod := range n.pods {
 		podKeys[i] = pod.Name
 	}
-	return fmt.Sprintf("&NodeInfo{Pods:%v, RequestedResource:%#v}", podKeys, n.requestedResource)
+	return fmt.Sprintf("&NodeInfo{Pods:%v, RequestedResource:%#v, NonZeroRequest: %#v}", podKeys, n.requestedResource, n.nonzeroRequest)
 }
 
 // addPod adds pod information to this NodeInfo.
 func (n *NodeInfo) addPod(pod *api.Pod) {
-	cpu, mem := calculateResource(pod)
+	cpu, mem, non0_cpu, non0_mem := calculateResource(pod)
 	n.requestedResource.MilliCPU += cpu
 	n.requestedResource.Memory += mem
+	n.nonzeroRequest.MilliCPU += non0_cpu
+	n.nonzeroRequest.Memory += non0_mem
 	n.pods = append(n.pods, pod)
 }
 
-func calculateResource(pod *api.Pod) (int64, int64) {
-	var cpu, mem int64
+// removePod subtracts pod information to this NodeInfo.
+func (n *NodeInfo) removePod(pod *api.Pod) error {
+	cpu, mem, non0_cpu, non0_mem := calculateResource(pod)
+	n.requestedResource.MilliCPU -= cpu
+	n.requestedResource.Memory -= mem
+	n.nonzeroRequest.MilliCPU -= non0_cpu
+	n.nonzeroRequest.Memory -= non0_mem
+
+	k1, err := getPodKey(pod)
```

Comment:
Please move it to the top - you don't want to change the underlying structure in that case, in case of error.